# Task 1

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('/content/adult.csv')
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [3]:
print(df.isnull().sum())

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


In [4]:
df.replace(' ?', np.nan, inplace=True)
print(df.isnull().sum())

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64


# Task 2

In [5]:
counts = df.isnull().sum()
percentages = (df.isnull().sum() / len(df)) * 100

missing_info = pd.DataFrame({
  'Missing Count': counts,
  'Missing Percentage': percentages
})
missing_info = missing_info.sort_values(by='Missing Percentage', ascending=False)
print(missing_info)


                Missing Count  Missing Percentage
age                         0                 0.0
workclass                   0                 0.0
fnlwgt                      0                 0.0
education                   0                 0.0
education.num               0                 0.0
marital.status              0                 0.0
occupation                  0                 0.0
relationship                0                 0.0
race                        0                 0.0
sex                         0                 0.0
capital.gain                0                 0.0
capital.loss                0                 0.0
hours.per.week              0                 0.0
native.country              0                 0.0
income                      0                 0.0


# Task 3

In [6]:
print(df.isnull().sum())

print("\nNo numeric columns with missing values were found so nothing to be dropped or imputed.")


age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

No numeric columns with missing values were found so nothing to be dropped or imputed.


# Task 4

In [7]:
duplicate_rows = df.duplicated().sum()
print(f"Duplicate rows: {duplicate_rows}")

df_attributes = df.drop(columns=['income'])
duplicates_without_income = df_attributes.duplicated().sum()

print(f"\nNumber of duplicate rows ignoring the 'income' column: {duplicates_without_income}")



Duplicate rows: 24

Number of duplicate rows ignoring the 'income' column: 25


# Task 5

In [8]:
print("\nEducation: ", df['education'].unique())
print("\nmarital.status: ", df['marital.status'].unique())
print("\nnative.country: ", df['native.country'].unique())



Education:  ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

marital.status:  ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

native.country:  ['United-States' '?' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


In [9]:
df['education'] = df['education'].str.strip()
df['marital.status'] = df['marital.status'].str.strip()
df['native.country'] = df['native.country'].str.strip()

print("\nEducation: ", df['education'].unique())
print("\nmarital.status: ", df['marital.status'].unique())
print("\nnative.country: ", df['native.country'].unique())


Education:  ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

marital.status:  ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

native.country:  ['United-States' '?' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


# Task 6

In [10]:
df.describe()

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [11]:
df.describe(include='object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


Mean age is 38 years with 40 hours per week wherease the median is 37 with 40 hours per week, so not much of a gap.

# Task 7

In [13]:
df["occupation"].value_counts()

,count
occupation,
Prof-specialty,4140
Craft-repair,4099
Exec-managerial,4066
Adm-clerical,3770
Sales,3650
Other-service,3295
Machine-op-inspct,2002
?,1843
Transport-moving,1597


In [16]:
df["sex"].value_counts(normalize=True) *100

,proportion
sex,
Male,66.920549
Female,33.079451


In [17]:
df["income"].value_counts(normalize=True) *100

,proportion
income,
<=50K,75.919044
>50K,24.080956


# Task 8

In [20]:
df.groupby("education")["education.num"].unique()

,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]


# Task 9

In [25]:
import pandas as pd
import sqlite3

# Read the raw CSV normally first
df_raw = pd.read_csv("adult.csv")

# Create/connect to a local SQLite database file
conn = sqlite3.connect("adult_income.db")

# Write the DataFrame into a SQL table called adult_income
df_raw.to_sql("adult_income", conn, if_exists="replace", index=False)

print("Table created. Original CSV shape:", df_raw.shape)

query = "SELECT * FROM adult_income WHERE age > 30;"
df_sql = pd.read_sql_query(query, conn)

print("Query result shape:", df_sql.shape)

# Cross-check: count rows manually with pandas filtering
expected = df_raw[df_raw["age"] > 30]
print("Pandas filter shape:", expected.shape)
print("SQL query shape:    ", df_sql.shape)

assert df_sql.shape == expected.shape, "Mismatch between SQL result and pandas filter!"
print("SQL extraction matches pandas filtering — extraction confirmed correct.")

conn.close()

Table created. Original CSV shape: (32561, 15)
Query result shape: (21989, 15)
Pandas filter shape: (21989, 15)
SQL query shape:     (21989, 15)
SQL extraction matches pandas filtering — extraction confirmed correct.


In [26]:
df_sql.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


# Task 10

32561 rows with 12 columns, dataset represents how adults have different properties with a certain income level, no missing values or incosistency, Prof-specialty is the most common occupation.

I didnt drop or impute since there was no missing values.